File: frontend/types/wiki.ts
export interface WikiPage {
  id: string
  course_id: string
  owner_id: string
  title: string
  slug: string
  summary: string | null
  content: string
  status: string
  current_version: number
  extra_meta: Record<string, unknown>
  created_at: string
  updated_at: string
  sources?: WikiSource[]
}

export interface WikiPageListItem {
  id: string
  course_id: string
  title: string
  slug: string
  summary: string | null
  status: string
  current_version: number
  created_at: string
  updated_at: string
}

export interface WikiPageVersion {
  id: string
  page_id: string
  version_number: number
  title: string
  content: string
  summary: string | null
  change_message: string | null
  created_by: string
  created_at: string
}

export interface WikiSource {
  id: string
  page_id: string
  source_type: string
  source_id: string
  source_title: string | null
  quote_text: string | null
  extra_meta?: Record<string, unknown>
  created_at?: string
}

export interface WikiLink {
  id: string
  source_page_id: string
  target_page_id: string
  relation_type: string
  created_at: string
}

export interface GenerateFromMaterialPayload {
  courseId: string
  materialId: string
}

export interface GenerateFromMaterialResponse {
  generated_count: number
  pages: { id: string; title: string; slug: string; current_version: number }[]
}

File: backend/app/schemas/wiki.py
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class WikiPageCreate(BaseModel):
    course_id: UUID
    title: str = Field(min_length=1, max_length=255)
    content: str = Field(min_length=1)
    summary: str | None = None


class WikiPageUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    content: str | None = None
    summary: str | None = None
    change_message: str | None = Field(default=None, max_length=500)


class WikiPageRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    course_id: UUID
    owner_id: UUID
    title: str
    slug: str
    summary: str | None = None
    content: str
    status: str
    current_version: int
    extra_meta: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class WikiPageListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    course_id: UUID
    owner_id: UUID
    title: str
    slug: str
    summary: str | None = None
    status: str
    current_version: int
    created_at: datetime
    updated_at: datetime


class WikiPageVersionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    page_id: UUID
    version_number: int
    title: str
    content: str
    summary: str | None = None
    change_message: str | None = None
    created_by: UUID
    created_at: datetime


class WikiPageListQuery(BaseModel):
    course_id: UUID
    status: str | None = "active"
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)


class WikiSourceRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    page_id: UUID
    source_type: str
    source_id: UUID
    source_title: str | None = None
    quote_text: str | None = None
    extra_meta: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


class WikiLinkRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    source_page_id: UUID
    target_page_id: UUID
    relation_type: str
    created_at: datetime


class GenerateFromMaterialRequest(BaseModel):
    course_id: UUID
    material_id: UUID


class UpdateFromNoteRequest(BaseModel):
    page_id: UUID
    note_content: str = Field(min_length=1)
